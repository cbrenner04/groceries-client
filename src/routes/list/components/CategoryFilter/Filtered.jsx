import React from 'react';
import PropTypes from 'prop-types';
import { Button } from 'react-bootstrap';

const Filtered = ({ filter, handleClearFilter }) => (
  <>
    <span id="filter-title">Filtering by:</span>
    <Button data-test-id="clear-filter" variant="outline-primary" id="filter-button" onClick={handleClearFilter}>
      {filter} <i className="fa fa-trash" />
    </Button>
  </>
);

Filtered.propTypes = {
  filter: PropTypes.string.isRequired,
  handleClearFilter: PropTypes.func.isRequired,
};

export default Filtered;
