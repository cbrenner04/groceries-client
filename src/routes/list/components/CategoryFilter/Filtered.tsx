import React, { type MouseEventHandler } from 'react';
import { Button } from 'react-bootstrap';

export interface IFilteredProps {
  filter: string;
  handleClearFilter: MouseEventHandler;
}

const Filtered: React.FC<IFilteredProps> = (props): React.JSX.Element => (
  <React.Fragment>
    <span id="filter-title">Filtering by:</span>
    <Button data-test-id="clear-filter" variant="outline-primary" id="filter-button" onClick={props.handleClearFilter}>
      {props.filter} <i className="fa fa-trash" />
    </Button>
  </React.Fragment>
);

export default Filtered;
