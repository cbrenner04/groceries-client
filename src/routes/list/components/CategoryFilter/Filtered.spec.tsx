import React from 'react';
import { render, type RenderResult } from '@testing-library/react';
import userEvent, { type UserEvent } from '@testing-library/user-event';

import Filtered, { type IFilteredProps } from './Filtered';

interface ISetupReturn extends RenderResult {
  user: UserEvent;
  props: IFilteredProps;
}

function setup(suppliedProps?: Partial<IFilteredProps>): ISetupReturn {
  const user = userEvent.setup();
  const defaultProps: IFilteredProps = {
    filter: 'Produce',
    handleClearFilter: jest.fn(),
  };
  const props = { ...defaultProps, ...suppliedProps };
  const component = render(<Filtered {...props} />);

  return { ...component, props, user };
}

describe('Filtered', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders with filter name in button', async () => {
    const { container, findByRole } = setup();

    expect(container).toMatchSnapshot();
    expect(await findByRole('button')).toHaveTextContent('Produce');
  });

  it('calls handleClearFilter when button is clicked', async () => {
    const { findByRole, props, user } = setup();

    await user.click(await findByRole('button'));

    expect(props.handleClearFilter).toHaveBeenCalled();
  });

  it('renders filter title text', async () => {
    const { findByText } = setup();

    expect(await findByText('Filtering by:')).toBeInTheDocument();
  });

  it('renders trash icon', async () => {
    const { container } = setup();

    const trashIcon = container.querySelector('.fa.fa-trash');
    expect(trashIcon).toBeInTheDocument();
  });

  it('has proper button styling', async () => {
    const { findByRole } = setup();

    const button = await findByRole('button');
    expect(button.className).toContain('btn');
    expect(button.className).toContain('btn-outline-primary');
  });

  it('has proper test ID for clear filter button', async () => {
    const { findByTestId } = setup();

    const button = await findByTestId('clear-filter');
    expect(button).toBeInTheDocument();
  });

  it('has proper ID for filter button', async () => {
    const { container } = setup();

    const button = container.querySelector('#filter-button');
    expect(button).toBeInTheDocument();
  });

  it('handles different filter names', async () => {
    const { findByRole } = setup({ filter: 'Dairy' });

    expect(await findByRole('button')).toHaveTextContent('Dairy');
  });

  it('handles filter names with special characters', async () => {
    const { findByRole } = setup({ filter: 'Frozen Foods & Desserts' });

    expect(await findByRole('button')).toHaveTextContent('Frozen Foods & Desserts');
  });

  it('handles uncategorized filter name', async () => {
    const { findByRole } = setup({ filter: 'uncategorized' });

    expect(await findByRole('button')).toHaveTextContent('Uncategorized');
  });

  it('handles long filter names', async () => {
    const longFilterName = 'Very Long Category Name That Might Wrap To Multiple Lines';
    const { findByRole } = setup({ filter: longFilterName });

    expect(await findByRole('button')).toHaveTextContent(longFilterName);
  });

  it('renders filter title with correct ID', async () => {
    const { container } = setup();

    const filterTitle = container.querySelector('#filter-title');
    expect(filterTitle).toBeInTheDocument();
    expect(filterTitle?.textContent).toBe('Filtering by:');
  });

  it('handles rapid button clicks', async () => {
    const { findByRole, props, user } = setup();

    const button = await findByRole('button');
    await user.click(button);
    await user.click(button);

    expect(props.handleClearFilter).toHaveBeenCalledTimes(2);
  });

  it('renders in a React Fragment', async () => {
    const { container } = setup();

    // The component should render multiple elements (span and button)
    const children = container.children;
    expect(children.length).toBeGreaterThan(0);
  });

  it('has proper button content structure', async () => {
    const { findByRole } = setup();

    const button = await findByRole('button');
    expect(button.textContent).toContain('Produce');
    expect(button.innerHTML).toContain('fa-trash');
  });
});
