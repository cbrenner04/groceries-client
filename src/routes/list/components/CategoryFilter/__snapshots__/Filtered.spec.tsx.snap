// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtered renders with filter name in button 1`] = `
<div>
  <span
    id="filter-title"
  >
    Filtering by:
  </span>
  <button
    class="btn btn-outline-primary"
    data-test-id="clear-filter"
    id="filter-button"
    type="button"
  >
    Produce
     
    <i
      class="fa fa-trash"
    />
  </button>
</div>
`;
